CenterForOpenScience/experimenter
# Do not allow login unless user has namespace permissions

Refs https://openscience.atlassian.net/browse/LEI-117
Companion to https://github.com/CenterForOpenScience/exp-addons/pull/47
## Purpose

The admin panel should not be open to all users. Those without namespace permissions should be explicitly prevented from logging in. (rather than letting them log in only to see error pages)

<img width="411" alt="screen shot 2016-03-09 at 10 25 42 pm" src="https://cloud.githubusercontent.com/assets/2957073/13658566/4ffbce72-e646-11e5-9d0e-b62e6e3b3ee4.png">
## Areas for feedback
- We should confirm that a failing request to the namespace is sufficient proof that the user lacks permissions
- Verify that this won't allow in other users, eg login-with-jam (if experimenter and lookit are to be merged into a single app)
- Error message improvements: I had some trouble getting the authenticator error to show the specific error message I intended (eg return value of the fail callback; it instead showed the return value of the parent request promise)
## Notes for testing

To test this, you will need to: 
1. Modify the bootstrap script; change occurrences of `user-osf-*` to `user-osf-<yourId>` for permission strings.
2. Rerun the bootstrap script.
3. Use a dummy OSF-staging account that does not match the ID above, and confirm that you are denied login with an error message like the one below.
- Confirm that any attempt to reach other pages takes you back to the login page
  1. Use your real OSF-staging account, and confirm that you are allowed to login and browse various pages.

If you need to flush your credentials, use `localStorage.clear()` at the console.
## Sample permissions script changes:

to bootstrap.js

```
diff --git a/scripts/bootstrap.js b/scripts/bootstrap.js
index 796c4cc..9a71d29 100644
--- a/scripts/bootstrap.js
+++ b/scripts/bootstrap.js
@@ -154,7 +154,7 @@ function setNamespacePermissions(token) {
     var ops = [
         {
             op: 'add',
-            path: '/permissions/user-osf-*',
+            path: '/permissions/user-osf-7bauz',
             value: 'ADMIN'
         }
     ];
@@ -177,7 +177,7 @@ function createThumbnailsCollection(token) {
         var ops = [
             {
                 op: 'add',
-                path: '/permissions/user-osf-*',
+                path: '/permissions/user-osf-7bauz',
                 value: 'ADMIN'
             },
             {
```

to permissions.js

```
diff --git a/scripts/permissions.js b/scripts/permissions.js
index c2ca42a..0527c0d 100644
--- a/scripts/permissions.js
+++ b/scripts/permissions.js
@@ -1,5 +1,5 @@
 // TODO make user specific
-var OSF = 'user-osf-*';
+var OSF = 'user-osf-7bauz';
 var PARTICIPANT = 'jam-experimenter:accounts-*';
 var GLOBAL = '*';

@@ -13,15 +13,15 @@ var admin = [
 ];

 var experiment = [
-    [GLOBAL, 'READ'],
+    //[GLOBAL, 'READ'],
     [OSF, 'CREATE']
 ];

-var session = [
-    [PARTICIPANT, 'CREATE'],
-    [PARTICIPANT, 'UPDATE'],
-    [OSF, 'READ']
-];
+//var session = [
+//    [PARTICIPANT, 'CREATE'],
+//    [PARTICIPANT, 'UPDATE'],
+//    [OSF, 'READ']
+//];

 var account = [
     [GLOBAL, 'CREATE'],

```
